sshd-common/src/main/java/org/apache/sshd/common/config/keys/OpenSshCertificateImpl.java

```diff
@@ -229,8 +228,81 @@ public String toString() {
                + "[id=" + getId()
                + ", serial=" + getSerial()
                + ", type=" + getType()
-               + ", validAfter=" + toDate(getValidAfter())
-               + ", validBefore=" + toDate(getValidBefore())
+               + ", validAfter=" + toStringDate(getValidAfter())
+               + ", validBefore=" + toStringDate(getValidBefore())
                + "]";
     }
+
+    @Override
+    public boolean equals(Object o) {
+        if (this == o) {
+            return true;
+        }
+        if (o == null || getClass() != o.getClass()) {
+            return false;
+        }
+
+        OpenSshCertificateImpl that = (OpenSshCertificateImpl) o;
+
+        if (serial != that.serial) {
+            return false;
+        }
+        if (type != that.type) {
+            return false;
+        }
+        if (keyType != null ? !keyType.equals(that.keyType) : that.keyType != null) {
+            return false;
+        }
+        if (!Arrays.equals(nonce, that.nonce)) {
+            return false;
+        }
+        if (certificatePublicKey != null
+                ? !certificatePublicKey.equals(that.certificatePublicKey) : that.certificatePublicKey != null) {
+            return false;
+        }
+        if (id != null ? !id.equals(that.id) : that.id != null) {
+            return false;
+        }
+        if (principals != null ? !principals.equals(that.principals) : that.principals != null) {
+            return false;
+        }
+        if (validAfter != null ? !validAfter.equals(that.validAfter) : that.validAfter != null) {
+            return false;
+        }
+        if (validBefore != null ? !validBefore.equals(that.validBefore) : that.validBefore != null) {
+            return false;
+        }
+        if (criticalOptions != null ? !criticalOptions.equals(that.criticalOptions) : that.criticalOptions != null) {
+            return false;
+        }
+        if (extensions != null ? !extensions.equals(that.extensions) : that.extensions != null) {
+            return false;
+        }
+        if (reserved != null ? !reserved.equals(that.reserved) : that.reserved != null) {
+            return false;
+        }
+        if (caPubKey != null ? !caPubKey.equals(that.caPubKey) : that.caPubKey != null) {
+            return false;
+        }
+        return Arrays.equals(signature, that.signature);
+    }
+
+    @Override
+    public int hashCode() {
```

Comment:
Again: hashCode() on non-final fields.

With non-final fields I think this should not have content-based equals() or hashCode(). As soon as clients can modify fields, this is asking for trouble.

Reply:
Removed equals/hashCode entirely from OpenSshCertificateImpl, this was only added to make equality checks in the test case simpler, but I just moved the equality checks to the unit test instead